pkg/policy/updaterequest.go

```diff
@@ -0,0 +1,160 @@
+package policy
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gardener/controller-manager-library/pkg/logger"
+	kyverno "github.com/kyverno/kyverno/api/kyverno/v1"
+	urkyverno "github.com/kyverno/kyverno/api/kyverno/v1beta1"
+	"github.com/kyverno/kyverno/pkg/config"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func (pc *PolicyController) updateUR(policyKey string, policy kyverno.PolicyInterface) {
+	logger := pc.log.WithName("updateUR").WithName(policyKey)
+	logger.Info("update URs on policy event")
+
+	mutateURs := pc.listMutateURs(policyKey, nil)
+	generateURs := pc.listGenerateURs(policyKey, nil)
+	updateUR(pc.kyvernoClient, policy.GetName(), append(mutateURs, generateURs...), pc.log.WithName("updateUR"))
+
+	for _, rule := range policy.GetSpec().Rules {
+		var ruleType urkyverno.RequestType
+		if rule.IsMutateExisting() {
+			ruleType = urkyverno.Mutate
+		} else {
+			// TODO: assign generate ruleType
+			continue
+		}
+
+		triggers := getTriggers(rule)
+		for _, trigger := range triggers {
+			var urs []*urkyverno.UpdateRequest
+			if ruleType == urkyverno.Mutate {
+				urs = pc.listMutateURs(policyKey, trigger)
+			} else {
+				urs = pc.listGenerateURs(policyKey, trigger)
+			}
+
+			logger.V(4).Info("UR was created", "rule", rule.Name, "rule type", ruleType, "trigger", trigger.Namespace+trigger.Name)
+
+			if urs != nil {
+				continue
+			}
+
+			logger.Info("creating new UR")
+			ur := newUR(policy, trigger, ruleType)
+			new, err := pc.kyvernoClient.KyvernoV1beta1().UpdateRequests(config.KyvernoNamespace).Create(context.TODO(), ur, metav1.CreateOptions{})
+			if err != nil {
+				pc.log.Error(err, "failed to create new UR policy update", "policy", policy.GetName(), "rule", rule.Name, "rule type", ruleType,
+					"target", fmt.Sprintf("%s/%s/%s/%s", trigger.APIVersion, trigger.Kind, trigger.Namespace, trigger.Name))
+				continue
+			} else {
+				pc.log.V(4).Info("successfully created UR on policy update", "policy", policy.GetName(), "rule", rule.Name, "rule type", ruleType,
+					"target", fmt.Sprintf("%s/%s/%s/%s", trigger.APIVersion, trigger.Kind, trigger.Namespace, trigger.Name))
+			}
+
+			new.Status.State = urkyverno.Pending
+			if _, err := pc.kyvernoClient.KyvernoV1beta1().UpdateRequests(config.KyvernoNamespace).UpdateStatus(context.TODO(), new, metav1.UpdateOptions{}); err != nil {
+				pc.log.Error(err, "failed to set UpdateRequest state to Pending")
```

Comment:
Don't we want to reconcile in case of err, should we return the err ?

Reply:
Yes, we continue on error to process the rest of the rules. This is the same as how it handles in the webhook.